include/zephyr/bluetooth/buf.h

```diff
@@ -193,26 +186,27 @@ struct net_buf *bt_buf_get_tx(enum bt_buf_type type, k_timeout_t timeout,
  */
 struct net_buf *bt_buf_get_evt(uint8_t evt, bool discardable, k_timeout_t timeout);
 
-/** Set the buffer type
+/** Set the buffer type. The type is encoded as an H:4 byte prefix as part of
+ *  the payload itself.
  *
  *  @param buf   Bluetooth buffer
  *  @param type  The BT_* type to set the buffer to
  */
-static inline void bt_buf_set_type(struct net_buf *buf, enum bt_buf_type type)
+static inline void bt_buf_set_type(struct net_buf *buf, uint8_t type)
 {
-	((struct bt_buf_data *)net_buf_user_data(buf))->type = type;
+	net_buf_push_u8(buf, type);
 }
 
-/** Get the buffer type
+/** Get the buffer type. This pulls the H:4 byte prefix from the payload, which means
+ *  that the call can be done only once per buffer.
  *
  *  @param buf   Bluetooth buffer
  *
  *  @return The BT_* type to of the buffer
  */
-static inline enum bt_buf_type bt_buf_get_type(struct net_buf *buf)
+static inline uint8_t bt_buf_get_type(struct net_buf *buf)
```

Comment:
I think we can and should deprecate `bt_buf_get_type` and `bt_buf_set_type`, and explicitly do a net_buf pull/push instead. Can we do it in this PR as well?

We need to keep these function for now in deprecated state to remain compatible with out-of-tree drivers. My understanding is that drivers will remain compatible so long as they only call these functions once per buffer, correct?

Reply:
> I think we can and should deprecate bt_buf_get_type and bt_buf_set_type, and explicitly do a net_buf pull/push instead. Can we do it in this PR as well?

Sure. I'm ok with that.

> My understanding is that drivers will remain compatible so long as they only call these functions once per buffer, correct?

Correct.

Reply:
Done